src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs

```diff
@@ -2445,6 +2445,12 @@ public sealed partial class TypeName
         public System.Reflection.Metadata.TypeName GetGenericTypeDefinition() { throw null; }
         public System.Reflection.Metadata.TypeName GetElementType() { throw null; }
         public int GetNodeCount() { throw null; }
+        public System.Reflection.Metadata.TypeName MakeArrayTypeName() { throw null; }
```

Comment:
Do you also plan to add `MakeSimpleTypeName` and `MakeNestedTypeName`?

Reply:
I think it would be better to expose those APIs and build `WithAssemblyName` as a helper in S.F.Nrbf. The behavior of WithAssemblyName is very Nrbf specific. I do not think it is generally usable method.

Reply:
> Do you also plan to add `MakeSimpleTypeName` and `MakeNestedTypeName`?

`MakeSimpleTypeName` would need to parse the name anyway. I think it's simpler to just use `TypeName.Parse`. Unless I am missing something (the method would throw for non-simple names?)?

`MakeNestedTypeName` I currently don't see any real-life usage for it and would prefer to wait for customer feedback about it.

Reply:
> I think it would be better to expose those APIs and build `WithAssemblyName` as a helper in S.F.Nrbf. 

But I don't know how to do that with the current APIs without the re-parsing.


> The behavior of `WithAssemblyName` is very Nrbf specific. I do not think it is generally usable method.

I've seen a need in WinForms but it's also related to BinaryFormatter:

https://github.com/dotnet/winforms/blob/31176a4d203711e766b8b587414d04c27d8cc173/src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs#L54-L61

I believe that having this method and for example allowing to pass null to clear all assembly names is generally usable.

Reply:
> I don't know how to do that with the current APIs without the re-parsing.

You just need to validate that the type name does not contain invalid escape sequence. It is much cheaper that full reparsing.

> I believe that having this method and for example allowing to pass null to clear all assembly names is generally usable.

I can believe that it is useful in the world of binary serializers that are not the primary design point of these APIs. I do not think it is useful for the world of ECMA and compilers.

Reply:
> You just need to validate that the type name does not contain invalid escape sequence. It is much cheaper/simpler than full reparsing.

Now I get the point, moreover I see the value and would use it myself in Nrbf. I've extended the proposal in https://github.com/dotnet/runtime/issues/102263. It seems that https://apireview.net/backlog is pretty packed, I will implement `static TypeName MakeSimpleTypeName` when I am back from my vacations and extend the proposal again.

> I can believe that it is useful in the world of binary serializers that are not the primary design point of these APIs. I do not think it is useful for the world of ECMA-335, compilers and similar tools.

And that API needs to serve both, so I don't see why we should not add it.

Reply:
> And that API needs to serve both, so I don't see why we should not add it.

`WithAssemblyName` is a nice-to-have convenience API for binary serializers. It is not required to make the binary serializers use case work. Binary serializers can have it implemented as a local helper just fine.

Reply:
> MakeNestedTypeName I currently don't see any real-life usage for it and would prefer to wait for customer feedback about it.

It is necessary for completeness so that one can implement arbitrary type name builder or rewriter (`WithAssemblyName` is an example of such rewriter).